python/ray/train/tests/test_xgboost_trainer.py

```diff
@@ -152,6 +145,46 @@ def test_checkpoint_freq(ray_start_4_cpus, freq_end_expected):
     assert cp_paths == sorted(cp_paths), str(cp_paths)
 
 
+@pytest.mark.parametrize("rank", [0, 1])
+def test_checkpoint_only_on_rank0(rank):
+    callback = RayTrainReportCallback(frequency=2, checkpoint_at_end=True)
+
+    booster = mock.MagicMock()
+
+    with mock.patch("ray.train.report") as mock_report, mock.patch(
+        "ray.train.get_context"
+    ) as mock_get_context:
+        mock_context = mock.MagicMock()
+        mock_context.get_world_rank.return_value = rank
+        mock_get_context.return_value = mock_context
+
+        booster.num_boosted_rounds.return_value = 2
+        callback.after_iteration(booster, epoch=1, evals_log={})
+
+        # Only rank 0 should report based on `frequency`
+        reported_checkpoint = bool(mock_report.call_args.kwargs.get("checkpoint"))
+        if rank == 0:
+            assert reported_checkpoint
+        else:
+            assert not reported_checkpoint
+
+        booster.num_boosted_rounds.return_value = 3
+        callback.after_iteration(booster, epoch=2, evals_log={})
+        # Nobody should report a checkpoint on iterations
+        reported_checkpoint = bool(mock_report.call_args.kwargs.get("checkpoint"))
+        assert not reported_checkpoint
```

Comment:
Is this test unnecessarily complicated? Any better ways to do this?

Reply:
Just went through some MagicMock docs. I think this unit-testing function looks good too me :)
For simplicity, probably we can just test the class variable `RayTrainReportCallback._last_checkpoint_iteration` after the function `callback.after_iteration()` is called. In this way, we can reduce the mock of `ray.train.report`. Then the code from line 154 ~ 169 could be simplified as:
```
with mock.patch("ray.train.get_context") as mock_get_context:
        mock_context = mock.MagicMock()
        mock_context.get_world_rank.return_value = rank
        mock_get_context.return_value = mock_context

        booster.num_boosted_rounds.return_value = 2
        callback.after_iteration(booster, epoch=1, evals_log={})
        assertEqual(callback._last_checkpoint_iteration, 1) # same as epoch number
```
However, this test assumes `ray.train.report` and `RayTrainReportCallback._get_checkpoint()` always works. It's also a little indirect. Correct me if I am wrong :)

Reply:
You're right, the test was a little too indirect and I was able to find a similar way to simplify it. PTAL

Reply:
Nice simplification! Learnt a lot from the code 👍